### PR TITLE
[bug fix] Need to wait for previous recover task finished. 

### PR DESCRIPTION
**Pull request by liuyu85cn**
## What type of PR is this?
- [X] bug
- [ ] feature
- [ ] enhancement

#### Issue(s) number: 

n/a

#### Description:

In TOSS, each time a raft part get leader, it will do a scan to see if there is any primes.
but if the leader changed really quick. There may be some previous term scan not finished. yet. 
These to recover task will write the same key and finished in any order. 

In this PR, the latest recover task will wait for previous task finish, then start. 

## Checklist:
Tests:
- [ ] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

n/a

### CODE REVIEW

File: src/storage/transaction/TransactionManager.cpp [hunk 1]
```diff
@@ -192,13 +194,16 @@ void TransactionManager::onLeaderLostWrapper(const ::nebula::kvstore::Part::Call
...
-  auto currTermIter = currTerm_.find(currTermKey);
-  if (currTermIter == currTerm_.end()) {
-    return;
+  // clean some out-dated item in memory lock
+  for (auto cit = memLocks_.cbegin(); cit != memLocks_.cend(); ++cit) {
+    auto& [spaceId, partId, termId] = cit->first;
+    if (spaceId == opt.spaceId && partId == opt.partId && termId < opt.term) {
+      auto sptrLockCore = cit->second;
+      if (sptrLockCore->size() == 0) {
+        cit = memLocks_.erase(cit);
```
Comment (panda-sheep) [inline src/storage/transaction/TransactionManager.cpp:203]: `cit` returned by erase and `++cit ` can cause missing elements

File: src/storage/transaction/TransactionManager.h [hunk 2]
```diff
@@ -129,12 +128,18 @@ class TransactionManager {
...
 
+  void waitUntil(std::function<bool()>&& cond, folly::Promise<folly::Unit>&& p);
+
  protected:
   using SpacePart = std::pair<GraphSpaceID, PartitionID>;
 
   StorageEnv* env_{nullptr};
-  std::shared_ptr<folly::IOThreadPoolExecutor> exec_;
+  std::shared_ptr<folly::IOThreadPoolExecutor> worker_;
+  std::shared_ptr<folly::IOThreadPoolExecutor> controller_;
```
Comment (panda-sheep) [inline src/storage/transaction/TransactionManager.h:138]: Can you add comments to these two variables?

Comment (liuyu85cn) [reply src/storage/transaction/TransactionManager.h:138]: ok

File: src/storage/transaction/TransactionManager.cpp (hunk 1 above)
Comment (liuyu85cn) [reply src/storage/transaction/TransactionManager.cpp:203]: good catch..

**Review by panda-sheep — APPROVED** [review]
Good job！

File: src/storage/transaction/TransactionManager.cpp [hunk 3]
```diff
@@ -211,54 +218,69 @@ void TransactionManager::onLeaderElectedWrapper(
...
-      auto succeed = lk->try_lock(edgeKey);
-      if (!succeed) {
-        LOG(ERROR) << "not supposed to lock fail: "
-                   << ", space " << spaceId << ", partId " << partId << ", termId " << termId
-                   << folly::hexlify(edgeKey);
+        if (!hasUnfinishedTask()) {
+          addPrime(spaceId, partId, termId, edgeKey, resumeVec[i]);
+        } else {
+          folly::Promise<folly::Unit> pro;
+          auto fut = pro.getFuture();
```
Comment (wenhaocs) [inline src/storage/transaction/TransactionManager.cpp:253]: minor: "pro" usually means good vs. con. It's better to use "p", "promise" or "xxPromise" to indicate promise.

**Review by wenhaocs — APPROVED** [review]
excellent

**Pull request merged**